### ISSUE TITLE
Provide mechanism to locally override node addresses

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#1821]
It's often desirable to use outbound pgbouncers with Citus. So far, one approach has been to use DNS trickery to direct all connections from postgres to a worker node into a pgbouncer running on localhost. However, logical replication connections cannot go through pgbouncer, and we therefore need a way to direct replication connections directly to postgres while directing queries through the pgbouncer.

One approach could be for Citus to read node address overrides from a user-defined variant of `/etc/hosts`, such that connections to other workers can be redirected to a local pgbouncer. A replication connection is opened by a postgres background process and therefore it would not have the override.

**Comment by fdr** [comment]
It's not just host names: don't forget about the port. We have a monstrous yet reliable gyration to make that work using loopback interfaces, so that we have the option of enabling or disabling pooling without messing with catalogs.

Perhaps Citus should learn to understand pooler addressing, much as it has learned the role of replication in the system.

In general, it should probably stop relying so much on `(host, port)` as the primary means of identification. Nodes should have some identifier does not interact with network and authentication addressing. It does need some care to make sure that the simplest configuration steps are not bloated, though.

**Comment by lithp** [comment]
> Nodes should have some identifier that does not interact with network and authentication addressing

[We have one of those](https://github.com/citusdata/citus/blob/master/src/backend/distributed/citus--6.0-3--6.0-4.sql#L16) :) Although I agree that more UDFs should accept it!

> One approach could be for Citus to read node address overrides from a user-defined variant of the /etc/hosts file, such that connections to other workers can be redirected to a local pgbouncer.

I think this makes sense, although don't know why we wouldn't put it into another metadata table / extra columns on pg_dist_node.

We could do:

```sql
CREATE TABLE pg_dist_node_bouncers(
  -- one entry for every worker node which has a bouncer in front of it
  nodeid int NOT NULL,

  -- the location user queries should be sent to
  bouncerport int NOT NULL,
  bouncername text NOT NULL
)
```

Or we could go the other direction; assume that `pg_dist_node.{nodename,nodeport}` identify the bouncers and add a table telling citus where the worker _actually_ is so it can correctly configure logical replication.

**Comment by jasonmp85** [comment]
> We could do […]

Would it make sense to go to a full connection string URL rather than name/port? I don't know if we need _every_ feature capable of configuration in a URL, but it feels like if we had started with that representation in the first place a lot more of the configuration could be both flexible and explicit in the URL for a worker.

**Comment by fdr** [comment]
I'm also nonplussed about not having "direct" libpq access via URI or DSN. That's where this whole "sslmode" stuff comes in, too. Or part of our inability to perform `verify-full` checking. The only issue I can think of there is confidentiality of certain sub-strings.

**Comment by jasonmp85** [comment]
Can you clarify your use of _nonplussed_? It has two contradictory meanings…

**Comment by fdr** [comment]
Hm. So it is. I also shouldn't use a double-negation.

I mean in the sense that I was surprised from the get-go that citus required an idiosyncratic way to define connection information that was not a direct `libpq` string or lightly pre/post-processed version (see: dblink). As you know, this was the root of a lot of machination, e.g. `.pgpass`.

Speaking of `.pgpass`, it may be reasonable at least look at obsoleting it.

**Comment by marcocitus** [comment]
Unlike most postgres features, Citus needs to differentiate  between different combinations of (hostname, port, user name and database name) for workers and localhost. We need to be able to map every possible combination to a connection string.

pgpass contains no more information than necessary to map (username, database name) combinations to a password. It seems like what we need here is primarily a way to map (hostname, port) to (hostname, port). It doesn't seem very practical to permute them into one giant mapping and specify other global configurations (e.g. SSL) for each tuple.

Maybe just a hostname:port -> hostname:port mapping in a text file?

**Comment by jasonmp85** [comment]
Wait what's this affinity for a text file? We had one and did a lot of work to make it go away…

**Comment by marcocitus** [comment]
> Wait what's this affinity for a text file? We had one and did a lot of work to make it go away…

pgbouncer set-up is specific to the host, not the database. The database gets replicated to various places (followers, standby). It doesn't seem like we would want to enforce the same config across all those hosts by storing it in the database. Also, there could be multiple databases that would all have to be configured separately.

**Comment by lithp** [comment]
> Also, there could be multiple databases that would all have to be
configured separately.

The counterpoint to this is that if you use a single file to configure all
of them there's going to be an implicit dependency between the
`pg_dist_node`'s of all the different databases.

Either you use `nodeid` and force that all nodes with the same `nodeid`,
globally, route to the same bouncer, or you use `(nodename, nodeport)` and
force that all traffic to the same destination go through the same bouncer.
Actually, that doesn't sound so bad, but you are giving up /some/
flexibility.

**Comment by marcocitus** [comment]
> Either you use `nodeid` and force that all nodes with the same `nodeid`,

I don't think you can use nodeid as a reliable long-lived node identifier since it changes if you remove and add a node, and depends on the order in which nodes are added.

**Comment by fdr** [comment]
Another related issue to try to consolidate gains on: https://github.com/citusdata/citus/issues/1047 bypassing the pooler for COPY (I'm seeing an interesting bug related to this right now).
